Generate EFI grub images for x86 platforms

sduponch opened the pull request:
- Add option in make menuconfig to build an EFI Grub image with legacy grub support.
- Update the x86 Makefile to generate an EFi grub image [ MBR | EFI | GrubModule | RootFs ]
- EFI image generate an UUID GPT signature to replace the too small legacy part signature.
- EFI image use the grub search files module to find the boot partition (as grub does not have a GPT uuid part search support).
- Fork grub build with efi platform support via a common.mk Makefile include to not fork the whole grub package and patches (host need both of legacy and efi modules to setup the bootloader on efi images).
- Add gptfdisk tools package to get sgdisk on host side, used for converting MBR to GPT tables.
- Fork gen_image_generic.sh in gen_image_efi.sh to add EFI related additional's partitions.
- Update platform.sh to support GPT UUID root signature in /proc/cmdline and use awk (that is already embeded in pivot root) to reorder bytes of the dumped partition signature).
- Create a kmod-efi-vars kernel module (modules.mk) to add EFI related Kernel config options that are required to boot with a console on HDMI (FB).
- Generate VDI and VMDK EFI images

Known limitations : 
- Sysupgrade -p is forced for EFI/GPT images.

```
GPT EFI image partition scheme
+--------------------------------------------------------------------------------------------------------------------------+
| MBR (raw 1MB) | Kernel (ext4 4MB) | EF00 (vfat 1MB) | EF02 (raw 1MB) | Rootfs (ext4/squash XX MB)  | GTP table (raw 16KB)|
+--------------------------------------------------------------------------------------------------------------------------+
```

Tested on : 
- Virtualbox legacy mode
- Virtualbox EFI mode
- Intel Nuc DE3815TYKE legacy mode
- Intel Nuc DE3815TYKE EFI mode
- Giada F-210 with an EFI only bios
- Intel Cherry trail platform with an EFI only bios

Not tested on x86 32bits platform

alive4ever reviewed (commented):
tools/gptfdisk/Makefile: This should be on a separate pull request, since it introduces a new tool / base package.

alive4ever commented on tools/gptfdisk/Makefile:
```diff
@@ -0,0 +1,36 @@
+#
+include $(TOPDIR)/rules.mk
+
+PKG_NAME:=gptfdisk
```
I mean this change and related changes, which introduces ``gdisk`` should be submitted as separate pull request.

sduponch replied:
Submit a separate pull request to add a tools that is not used (yet) ? 
I can make a separate commit only for the tools, but i don't think the PR will be merged if the tool is not used...

alive4ever commented on target/linux/x86/image/Makefile:
```diff
@@ -10,6 +10,9 @@ include $(INCLUDE_DIR)/image.mk
 GRUB2_MODULES = biosdisk boot chain configfile ext2 linux ls part_msdos reboot serial vga
+GRUB2_MODULES_LEGACY = $(GRUB2_MODULES)
+GRUB2_MODULES_LEGACY += part_gpt part_msdos search fat exfat ext2
+GRUB2_MODULES_EFI = boot chain configfile ext2 linux ls part_msdos reboot serial part_gpt part_msdos search fat exfat ext2 efi_gop efi_uga gfxterm
```
There is no font module here. Does the grub framebuffer work, i.e. displays kernel text message?

sduponch replied:
Good catch, the kernel messages are not displayed until the system load the framebuffer module.

sduponch replied:
I am not sure if grub font are really required, i did not see any difference ... ?

alive4ever commented on target/linux/x86/image/Makefile:
```diff
@@ -217,14 +325,22 @@ define Image/Build
 	$(call Image/Build/$(1))
   ifneq ($(1),iso)
 	$(call Image/Build/grub2,$(1))
+	$(call Image/Build/efi,$(1))
```
This doesn't depend on ``ifeq ($(1),ext4)``. This means the UEFI images will be generated for both ext4 and squashfs image.

Does the generated squashfs image boot on EFI system?

I put the conditional ext4 so that the UEFI image is generated only for ext4 filesystem since I can't get the squashfs image to boot on EFI system.

sduponch replied:
Yes, the squashfs image boot correctly with EFI because the kernel partition is in ext4, only the rootfs partition is in squashfs (because squashfs is read only with an overlay in jffs2).

alive4ever commented on package/boot/grub2/grub2-efi/Makefile:
```diff
@@ -0,0 +1,46 @@
+  SUBMENU:=$(OTHER_MENU)
+  TITLE:=EFI VARS modules
+  DEPENDS:=@EFI_IMAGES
+  KCONFIG:= \
```
These kernel configurations should be included in the subtarget config (``target/linux/x86/64/config-default`` and ``target/linux/x86/generic/config-default``).

To modify the kernel configuration for subtarget, launch ``make CONFIG_TARGET=subtarget kernel_menuconfig``.

sduponch replied:
I think, the best place to do this is in target/linux/x86/modules.mk . We don't really need to modify the config-default to only support EFI and we don't want to append EFI stuff in a non EFI system.

alive4ever commented on tools/gptfdisk/patches/000-gdisk-only.patch:
```diff
@@ -0,0 +1,13 @@
+ LIB_HEADERS=$(LIB_NAMES:=.h)
+ DEPEND= makedepend $(CXXFLAGS)
+ 
+-all:	cgdisk gdisk sgdisk fixparts
```
This doesn't matter at all, since ``gdisk`` is host only package.

As I know, ``gdisk`` is intended for interactive use. For scripting and automation, ``sgdisk`` should be used.

sduponch replied:
i did not success to use sgdisk :\

alive4ever replied:
I added ``sgdisk`` equivalent for those actions.

alive4ever commented on config/Config-images.in:
```diff
@@ -224,6 +224,14 @@ menu "Target Images"
+		depends on TARGET_x86_64 || (TARGET_x86 && !TARGET_x86_rdc)
+		depends on TARGET_ROOTFS_EXT4FS || TARGET_ROOTFS_ISO || TARGET_ROOTFS_JFFS2 || TARGET_ROOTFS_SQUASHFS
+		select GRUB_IMAGES
+		select PACKAGE_grub2-efi
```
No need to introduce a new package here.

sduponch replied:
Do we need an EFI platform grub package on EFI target (this package is not for the host side) ?

alive4ever commented on package/boot/grub2/grub2-efi/Makefile:
```diff
@@ -0,0 +1,46 @@
+include $(TOPDIR)/rules.mk
+include $(INCLUDE_DIR)/kernel.mk
+
+PKG_NAME:=grub-efi
```
No need for new package here.

sduponch replied:
A separate package could be avoided if build variant were supported for Host packages :\

So, i made a common.mk Makefile inspired from the toolchain/gcc to ease the maintainability of these grub packages. I will take a look on how you did that :)

alive4ever reviewed (changes requested):
Here are my review, since the changes are rather too big for a single commit.

alive4ever commented on target/linux/x86/image/Makefile:
```diff
@@ -66,6 +70,92 @@ ifneq ($(CONFIG_GRUB_IMAGES),)
+	dd if="/dev/zero" of="$(BIN_DIR)/$(IMG_PREFIX)-uefi-gpt-$(1).img" bs=512 count=33 oflag=append conv=notrunc
+
+	# Set the new EFI partition (data are already present on image) and change PARTUUID of the boot partition
+	(echo n; echo 3; echo ; echo ; echo EF00; echo x; echo c; echo 2; echo $(EFI_SIGNATURE); echo w; echo Y) | gdisk "$(BIN_DIR)/$(IMG_PREFIX)-uefi-gpt-$(1).img"
```
sgdisk equivalent

Creating partition 3 with type EF00 (EFI System Partition).
``sgdisk -n 3 -t 3:EF00 "$(BIN_DIR)/$(IMG_PREFIX)-uefi-gpt-$(1).img"``

Changing UUID of partition 3 with predefined value.
``sgdisk -u 3:$(EFI_SIGNATURE) "$(BIN_DIR)/$(IMG_PREFIX)-uefi-gpt-$(1).img"``

alive4ever commented on target/linux/x86/image/Makefile:
```diff
@@ -66,6 +70,92 @@ ifneq ($(CONFIG_GRUB_IMAGES),)
+
+	# Append Grub EFI modules partition
+	dd if=/dev/zero of="$(BIN_DIR)/$(IMG_PREFIX)-uefi-gpt-$(1).img" bs=512 count=2048 oflag=append conv=notrunc
+	(echo x; echo e; echo m; echo n; echo 4; echo ; echo ; echo EF02; echo w; echo Y; echo Y) | gdisk "$(BIN_DIR)/$(IMG_PREFIX)-uefi-gpt-$(1).img"
```
sgdisk equivalent

Moving second header (backup mbr) to the end of disk.
``sgdisk -e "$(BIN_DIR)/$(IMG_PREFIX)-uefi-gpt-$(1).img"``

Creating partition 4 as bios boot partition
``sgdisk -n 4 -t 4:EF02 "$(BIN_DIR)/$(IMG_PREFIX)-uefi-gpt-$(1).img"``

sduponch replied:
Use of sgdisk add libpopt-dev as host dependency.
 sgdisk -n 3 -t 3:EF00 "$(BIN_DIR)/$(IMG_PREFIX)-uefi-gpt-$(1).img"
require use of the -g parameter to create GPT table

sduponch replied:
- sgdisk -u 3:$(EFI_SIGNATURE) "$(BIN_DIR)/$(IMG_PREFIX)-uefi-gpt-$(1).img"
+ sgdisk -u 2:$(EFI_SIGNATURE) "$(BIN_DIR)/$(IMG_PREFIX)-uefi-gpt-$(1).img"
GPT Signature must be on rootfs partition

alive4ever reviewed (commented):
I'll run a build test to check whether this single commit works as expected.

alive4ever commented on target/linux/x86/image/Makefile:
```diff
@@ -66,6 +70,92 @@ ifneq ($(CONFIG_GRUB_IMAGES),)
+	$(CP) $(STAGING_DIR_HOST)/lib/grub/i386-pc/*.img $(KDIR)/grub2/
+
+	# In order to append the EFI partitions we must pad the image so TARGET_IMAGES_PAD is forced
+	PADDING="y" SIGNATURE="$(SIGNATURE)" PATH="$(TARGET_PATH)" ./gen_image_generic.sh \
```
Instead of forcing ``y`` as value, I think it's better to make use of kconfig to select ``CONFIG_TARGET_IMAGES_PAD`` as ``EFI_IMAGES`` dependency.

sduponch replied:
We need to force image PADDING here, because we append new partitions to the image if we set in CONFIG others images will be padded too.

alive4ever commented on config/Config-images.in:
```diff
@@ -224,6 +224,14 @@ menu "Target Images"
+		bool "Build EFI GRUB images (Linux x86 or x86_64 host only)"
+		depends on TARGET_x86_64 || (TARGET_x86 && !TARGET_x86_rdc)
+		depends on TARGET_ROOTFS_EXT4FS || TARGET_ROOTFS_ISO || TARGET_ROOTFS_JFFS2 || TARGET_ROOTFS_SQUASHFS
+		select GRUB_IMAGES
```
On ``target/linux/x86/image/Makefile``, there is a statement that ``TARGET_IMAGES_PAD`` is forced.

``TARGET_IMAGES_PAD`` can be selected here instead of forcing ``TARGET_IMAGES_PAD=y``

sduponch replied:
Selecting TARGET_IMAGES_PAD in CONFIG will affect size of the generated non the EFI images...

sduponch replied:
Thanks a lot for your review.

I have updated the Makefiles to use sgdisk and moved the kmod efi-vars from grub2-efi to modules.mk.

sduponch replied:
PR splitted in two commit : 
- For sgdisk tools : https://github.com/openwrt/openwrt/pull/391/commits/3871e7c5c1e5e3beff8127e45d7d2f70dff174d7
- For EFI Grub : https://github.com/openwrt/openwrt/pull/391/commits/a9b15e43862459f45c75edf10aac5f82e150043a

alive4ever reviewed (commented):
Why not use kconfig to define ``TARGET_IMAGES_PAD`` instead of forcing ``TARGET_IMAGES_PAD=y``?

sduponch committed:
Add sgdisk in tools to be able to convert MBR to GPT

Signed-off-by: DUPONCHEEL Sébastien <sebastien.duponcheel@corp.ovh.com>

sduponch committed:
Generate EFI grub images for x86 platforms

Signed-off-by: DUPONCHEEL Sébastien <sebastien.duponcheel@corp.ovh.com>

alive4ever commented on target/linux/x86/image/Makefile:
```diff
@@ -66,6 +70,94 @@ ifneq ($(CONFIG_GRUB_IMAGES),)
+	rm $(KDIR)/kernel.efi || true
+	mkfs.fat -C $(KDIR)/kernel.efi -S 512 $(shell echo $$(( $(CONFIG_TARGET_KERNEL_PARTSIZE) * 1024 )))
+	mcopy -s -i "$(KDIR)/kernel.efi" $(KDIR)/grub2.efi/* ::/
+
```
How about moving this step to replace the first boot partition generation process, so that the first partition is EFI system partition (EF00), the second partition is bios boot (EF02), and the third partition is the root partition?.

This is important, because there will be a lot of space at the end of the disk after performing ``dd`` of the grub-efi-gpt image. With root partition placed on the last block, it is possible to expand the root partition to fill up entire free space without messing up another partition (EFI and bios boot).

alive4ever commented on target/linux/x86/image/Makefile:
```diff
@@ -66,6 +70,94 @@ ifneq ($(CONFIG_GRUB_IMAGES),)
+	# Append Grub EFI modules partition
+	dd if=/dev/zero of="$(BIN_DIR)/$(IMG_PREFIX)-uefi-gpt-$(1).img" bs=512 count=2048 oflag=append conv=notrunc
+	sgdisk -e "$(BIN_DIR)/$(IMG_PREFIX)-uefi-gpt-$(1).img"
+	sgdisk -n 4 -t 4:EF02 "$(BIN_DIR)/$(IMG_PREFIX)-uefi-gpt-$(1).img"
```
The BIOS boot partition should be placed before the root partition, so it's possible to expand the root partition to fill entire free space.

```
+-----------------------------+----------------------------+----------------+
| EFI System Partition (EF00) | BIOS boot partition (EF02) | Root partition |
+-----------------------------+----------------------------+----------------+
```

sduponch replied:
Why not, we can try this :)

sduponch replied:
i have started to work here: 
- https://github.com/sduponch/openwrt/pull/2/files

For now, the image build and grub boot in legacy and efi mode but system fail to mount rootfs because of a bad geometry (block count exceeds size of device)

alive4ever reviewed (commented):
How about placing the root partition at the end so that it's possible to expand the root partition without messing with another partition afterwards?

sduponch commented:
> How about placing the root partition at the end so that it's possible to expand the root partition without messing with another partition afterwards?
Done :)

alive4ever commented on target/linux/x86/image/Makefile:
```diff
@@ -10,6 +10,9 @@ include $(INCLUDE_DIR)/image.mk
 
 GRUB2_MODULES = biosdisk boot chain configfile ext2 linux ls part_msdos reboot serial vga
+GRUB2_MODULES_LEGACY = $(GRUB2_MODULES)
+GRUB2_MODULES_LEGACY += part_gpt part_msdos search fat exfat ext2
```
``part_msdos`` and ``ext2`` mentioned twice here.

alive4ever reviewed (commented):
Duplicate modules mentioned on ``GRUB2_MODULES_LEGACY`` variable

sduponch commented:
Done. I have also changed "gen_image_efi.sh" to output MBR image and be able to setup the legacy grub before converting this partition to GPT (should improve hybrid compatibility)

sduponch commented:
+---------------------+-----------------------+----------------------+--------------------- +-----------------------------+
| MBR (raw 1MB) | Kernel (ext4 1MB) | EF00 (vfat 1MB) | EF02 (raw 1MB) | Rootfs (ext/sq XX MB)  |
+---------------------+-----------------------+----------------------+----------------------+-----------------------------+

alive4ever commented:
So fast!

That's all my review.

Let's hope that OpenWRT/LEDE developers got the guts to decide whether to commit the changes or discard these altogether, so that we can move on to another task.

alive4ever commented on package/boot/grub2/common.mk:
```diff
@@ -1,33 +1,24 @@
 	http://gnualpha.uib.no/grub/ \
 	http://mirrors.fe.up.pt/pub/gnu-alpha/grub/ \
 	http://www.nic.funet.fi/pub/gnu/alpha/gnu/grub/
 PKG_MD5SUM:=be62932eade308a364ea4bbc91295930
```
One more forgotten review if you don't mind.

I suggest replacing ``PKG_MD5SUM`` with ``PKG_HASH``.
``PKG_HASH`` is just SHA256 hash, which should be very easy to generate.

sduponch replied:
Done

alive4ever commented on tools/gptfdisk/Makefile:
```diff
@@ -0,0 +1,36 @@
+
+PKG_SOURCE:=$(PKG_NAME)-$(PKG_VERSION).tar.gz
+PKG_SOURCE_URL:=http://www.rodsbooks.com/gdisk/
+PKG_MD5SUM:=d7f3d306b083123bcc6f5941efade586
```
This section should also be converted to ``PKG_HASH``

sduponch replied:
Done

sduponch commented:
I have fixed the grub2-efi "--with-suffix" warning (the option is --program-suffix) and updated Makefile.
This allow to have no footprint on MBR image because we fully separate the two grub binaries.

sduponch commented:
Backported to LEDE: https://github.com/lede-project/source/pull/1026

sduponch commented:
Closing because outdated by the LEDE port
